File: mafbase_stream/android/src/main/kotlin/com/example/mafbase_stream/pipeline/StreamingController.kt
package com.example.mafbase_stream.pipeline

import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.util.Log
import android.util.Size
import android.view.Surface
import com.example.mafbase_stream.StreamSession
import com.example.mafbase_stream.encoder.AudioPipeline
import com.example.mafbase_stream.events.StreamEventBus
import com.example.mafbase_stream.gl.Compositor
import com.example.mafbase_stream.jni.StreamSessionNative

/**
 * RTMP-стрим: [StreamSession] как выход компоситора плюс супервизор, который пересоздаёт
 * сессию после фатальной ошибки с backoff, пока стрим не остановил пользователь.
 *
 * Публичные методы — с main-потока, колбэки [Host] — на нём же. Старт и стоп сессии
 * (MediaCodec, AudioRecord, RTMP) блокирующие и выполняются в фоновых потоках.
 */
internal class StreamingController(
    private val config: StreamPipeline.Config,
    private val audioPipeline: AudioPipeline,
    private val compositorHost: CompositorHost,
    private val host: Host,
) {

    interface Host {
        fun isReleased(): Boolean

        /** Общий переход пайплайна: перезапуск стрима ждёт ролловер записи. */
        fun isTransitioning(): Boolean

        fun frameSize(): Size?

        fun videoBitrateBps(): Int

        fun onStateChanged()

        fun onActiveChanged()

        fun onMessage(text: String, long: Boolean)
    }

    private val mainHandler = Handler(Looper.getMainLooper())
    private var session: StreamSession? = null

    /**
     * Стрим включён пользователем. Остаётся true и пока сессия пересоздаётся после
     * ошибки — сбрасывается только по [stop] или провалу первого старта.
     */
    var isStreaming: Boolean = false
        private set

    var isStreamTransition: Boolean = false
        private set

    private var restartRunnable: Runnable? = null
    private val restartBackoff = RetryBackoff(baseMs = RESTART_BASE_MS, capMs = RESTART_CAP_MS)
    private var lastErrorMessageMs: Long = 0L
    private var stopRequested: Boolean = false

    fun start() {
        if (isStreaming || isStreamTransition) return
        val size = host.frameSize() ?: return
        if (!compositorHost.isCreated) return
        isStreamTransition = true
        stopRequested = false
        restartBackoff.reset()
        host.onStateChanged()

        val fresh = createSession(size, isRestart = false)
        Thread({ startBlocking(fresh, isRestart = false) }, "StreamSession-start").start()
    }

    fun stop() {
        cancelRestart()
        val current = session
        if (current == null) {
            if (isStreamTransition) {
                // Сессия сейчас стартует или пересоздаётся — её завершение подхватит флаг.
                stopRequested = true
            } else if (isStreaming) {
                // Ждали backoff перед пересозданием — сессии нет, просто гасим намерение.
                isStreaming = false
                host.onActiveChanged()
                host.onStateChanged()
            }
            return
        }
        session = null
        isStreamTransition = true
        host.onStateChanged()

        // Сначала отцепляем encoder Surface от Compositor'а ДО session.stop() —
        // иначе compositor продолжит eglSwapBuffers на разрушенный BufferQueue.
        compositorHost.detachOutput(Compositor.OutputId.STREAM_ENCODER)

        stopAsync(current) {
            isStreaming = false
            isStreamTransition = false
            host.onActiveChanged()
            host.onStateChanged()
            host.onMessage("Стрим остановлен", long = false)
        }
    }

    /** Вариант для освобождения пайплайна: без UI, стартующую сессию догонит [onSessionStarted]. */
    fun stopForRelease() {
        isStreaming = false
        val current = session ?: return
        session = null
        compositorHost.detachOutput(Compositor.OutputId.STREAM_ENCODER)
        stopAsync(current, onDone = null)
    }

    fun cancelRestart() {
        restartRunnable?.let { mainHandler.removeCallbacks(it) }
        restartRunnable = null
    }

    private fun createSession(size: Size, isRestart: Boolean): StreamSession {
        val fullUrl = if (config.streamKey.isNotBlank()) {
            "${config.rtmpUrl.trimEnd('/')}/${config.streamKey}"
        } else {
            config.rtmpUrl
        }
        val fresh = StreamSession(
            StreamSession.Config(
                rtmpUrl = fullUrl,
                width = size.width,
                height = size.height,
                videoBitrate = host.videoBitrateBps(),
            ),
            audioPipeline = audioPipeline,
        )
        fresh.setListener(object : StreamSession.Listener {
            override fun onStreamStarted() {
                mainHandler.post {
                    if (session !== fresh) return@post
                    restartBackoff.reset()
                    host.onMessage(if (isRestart) "Стрим восстановлен" else "Стрим запущен", long = false)
                }
            }

            override fun onStreamError(t: Throwable) {
                Log.e(TAG, "stream error", t)
                mainHandler.post { onSessionFailed(fresh, t.message ?: t.javaClass.simpleName) }
            }

            override fun onStreamEvent(event: StreamSessionNative.StreamEvent) {
                StreamEventBus.emit(event)
                if (event.type == StreamSessionNative.EventType.Failed) {
                    mainHandler.post { onSessionFailed(fresh, event.reason ?: "reconnect exhausted") }
                }
            }
        })
        return fresh
    }

    /**
     * Тело фонового потока старта: StreamSession.start() выполняет MediaCodec
     * prepare/configure/createInputSurface и AudioRecord init — на устройствах это
     * занимает 200-500 мс, на UI-потоке Choreographer писал бы «Skipped 41 frames!».
     */
    private fun startBlocking(fresh: StreamSession, isRestart: Boolean) {
        val ok = try {
            fresh.start()
            true
        } catch (e: Exception) {
            Log.e(TAG, "StreamSession.start failed", e)
            try {
                fresh.stop()
            } catch (_: Throwable) {
            }
            mainHandler.post { onSessionStartFailed(isRestart, e.message) }
            false
        }
        if (!ok) return

        val encoderSurface = fresh.encoderSurface
        if (encoderSurface == null) {
            Log.e(TAG, "stream encoder surface is null")
            fresh.stop()
            mainHandler.post { onSessionStartFailed(isRestart, null) }
            return
        }

        mainHandler.post { onSessionStarted(fresh, encoderSurface) }
    }

    private fun onSessionStartFailed(isRestart: Boolean, message: String?) {
        isStreamTransition = false
        if (isRestart && isStreaming && !stopRequested && !host.isReleased()) {
            scheduleRestart()
        } else {
            isStreaming = false
            stopRequested = false
            host.onActiveChanged()
            if (!isRestart) host.onMessage("Не удалось начать стрим: $message", long = true)
        }
        host.onStateChanged()
    }

    private fun onSessionStarted(fresh: StreamSession, encoderSurface: Surface) {
        if (host.isReleased() || stopRequested || !compositorHost.isCreated) {
            // Пока сессия поднималась, стрим остановили — гасим её, не подключая.
            stopAsync(fresh) {
                isStreamTransition = false
                isStreaming = false
                stopRequested = false
                host.onActiveChanged()
                host.onStateChanged()
            }
            return
        }
        session = fresh
        compositorHost.attachOutput(Compositor.OutputId.STREAM_ENCODER, encoderSurface, needsPresentationTime = true)
        val wasStreaming = isStreaming
        isStreaming = true
        isStreamTransition = false
        if (!wasStreaming) host.onActiveChanged()
        host.onStateChanged()
    }

    private fun stopAsync(target: StreamSession, onDone: (() -> Unit)?) {
        Thread({
            try {
                target.stop()
            } catch (e: Exception) {
                Log.e(TAG, "StreamSession.stop failed", e)
            }
            if (onDone != null) mainHandler.post(onDone)
        }, "StreamSession-stop").start()
    }

    /**
     * Сессия умерла (ядро прислало Failed или энкодер выдал ошибку) — пересоздаём её
     * с backoff, пока стрим не остановил пользователь.
     */
    private fun onSessionFailed(failed: StreamSession, reason: String) {
        if (host.isReleased() || session !== failed || !isStreaming) return
        val now = SystemClock.elapsedRealtime()
        if (now - lastErrorMessageMs >= ERROR_MESSAGE_INTERVAL_MS) {
            lastErrorMessageMs = now
            host.onMessage("Ошибка стрима: $reason. Переподключаемся…", long = true)
        }
        scheduleRestart()
    }

    /** Backoff 2→4→8→15 с без лимита попыток. */
    private fun scheduleRestart() {
        if (host.isReleased() || restartRunnable != null) return
        val delayMs = restartBackoff.nextDelayMs()
        Log.w(TAG, "stream restart #${restartBackoff.attempt} in ${delayMs}ms")
        postRestart(delayMs)
    }

    private fun postRestart(delayMs: Long) {
        val runnable = Runnable {
            restartRunnable = null
            restartNow()
        }
        restartRunnable = runnable
        mainHandler.postDelayed(runnable, delayMs)
    }

    private fun restartNow() {
        if (host.isReleased() || !isStreaming) return
        if (host.isTransitioning()) {
            // Идёт ролловер сегмента записи или другой переход — не мешаем, пробуем позже.
            postRestart(TRANSITION_RETRY_MS)
            return
        }
        val size = host.frameSize() ?: return
        if (!compositorHost.isCreated) return
        val old = session
        session = null
        isStreamTransition = true
        host.onStateChanged()
        compositorHost.detachOutput(Compositor.OutputId.STREAM_ENCODER)

        val fresh = createSession(size, isRestart = true)
        Thread({
            if (old != null) {
                try {
                    old.stop()
                } catch (e: Exception) {
                    Log.e(TAG, "StreamSession.stop (restart) failed", e)
                }
            }
            startBlocking(fresh, isRestart = true)
        }, "StreamSession-restart").start()
    }

    companion object {
        private const val TAG = "StreamingController"
        private const val RESTART_BASE_MS = 2_000L
        private const val RESTART_CAP_MS = 15_000L
        private const val ERROR_MESSAGE_INTERVAL_MS = 60_000L
    }
}

File: mafbase_stream/ios/Classes/Pipeline/StreamingController.swift
import Foundation
import Network

/// RTMP-стрим: `StreamSession` как выход компоситора плюс супервизор, который пересоздаёт
/// сессию после сбоя с backoff, пока стрим не остановил пользователь, а при смене сетевого
/// пути делает это сразу, не дожидаясь backoff.
///
/// Публичные методы и колбэки — на main queue. Старт и стоп сессии блокирующие и идут на
/// фоновой очереди; `session` читают очереди кадров и звука.
final class StreamingController {

    private static let restartBaseMs = 2000
    private static let restartCapMs = 15000
    /// Пауза перед повтором рестарта, которому помешал ролловер записи.
    private static let transitionRetry: TimeInterval = 1
    private static let errorMessageInterval: TimeInterval = 60

    /// Новая сессия с текущими параметрами пайплайна; `nil`, если пайплайна уже нет.
    var makeSession: (() -> StreamSession?)?
    var isReleased: (() -> Bool)?
    /// Рестарт стрима ждёт ролловер записи.
    var isRecordTransition: (() -> Bool)?
    var onStateChanged: (() -> Void)?
    var onMessage: ((String, Bool) -> Void)?

    /// Подключённая сессия — выход для кадров и звука; `nil`, пока сессия стартует или
    /// ждёт пересоздания.
    private(set) var session: StreamSession?
    /// Стрим включён пользователем. Остаётся true, пока сессия пересоздаётся после сбоя;
    /// сбрасывается по `stop()` или провалу самого первого старта.
    private(set) var isStreaming = false
    private(set) var isStreamTransition = false

    private let restartBackoff = RetryBackoff(
        baseMs: StreamingController.restartBaseMs,
        capMs: StreamingController.restartCapMs
    )
    private var restartTimer: Timer?
    private var lastErrorMessageAt: Date?
    /// Стоп пришёл, пока сессия стартовала — её погасит завершение старта.
    private var stopRequested = false
    private var pathMonitor: NWPathMonitor?
    private var lastPathKey: String?

    // MARK: - Controls

    func start() {
        guard !isStreaming, !isStreamTransition, let fresh = makeSession.flatMap({ $0() }) else { return }
        isStreamTransition = true
        stopRequested = false
        restartBackoff.reset()
        onStateChanged?()
        startPathMonitor()
        launch(fresh, replacing: nil, isRestart: false)
    }

    func stop() {
        cancelRestart()
        stopPathMonitor()
        guard let current = session else {
            if isStreamTransition {
                stopRequested = true
            } else if isStreaming {
                isStreaming = false
                onStateChanged?()
            }
            return
        }
        session = nil
        isStreaming = false
        isStreamTransition = true
        onStateChanged?()

        StreamPipeline.runProtectedFromSuspension(name: "mafbase_stream.stop-stream") { done in
            DispatchQueue.global(qos: .userInitiated).async { [weak self] in
                current.stop()
                DispatchQueue.main.async {
                    done()
                    guard let self = self else { return }
                    self.isStreamTransition = false
                    self.onStateChanged?()
                }
            }
        }
    }

    /// Для освобождения пайплайна: без UI, стартующую сессию погасит завершение её старта.
    func stopForRelease() {
        cancelRestart()
        stopPathMonitor()
        isStreaming = false
        if isStreamTransition {
            stopRequested = true
        }
        guard let current = session else { return }
        session = nil
        StreamPipeline.runProtectedFromSuspension(name: "mafbase_stream.stop-stream") { done in
            current.stop()
            done()
        }
    }

    // MARK: - Session lifecycle

    private func launch(_ fresh: StreamSession, replacing old: StreamSession?, isRestart: Bool) {
        bind(fresh, isRestart: isRestart)
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            old?.stop()
            do {
                try fresh.start()
            } catch {
                NSLog("[mafbase_stream] StreamSession.start failed: \(error)")
                fresh.stop()
                DispatchQueue.main.async { self?.handleStartFailed(isRestart: isRestart, error: error) }
                return
            }
            DispatchQueue.main.async { self?.handleStarted(fresh) }
        }
    }

    private func bind(_ fresh: StreamSession, isRestart: Bool) {
        fresh.onStarted = { [weak self] in
            DispatchQueue.main.async {
                guard let self = self, self.session === fresh else { return }
                self.restartBackoff.reset()
                self.onMessage?(isRestart ? "Стрим восстановлен" : "Стрим запущен", false)
            }
        }
        fresh.onError = { [weak self] error in
            DispatchQueue.main.async { self?.handleFailure(of: fresh, reason: "\(error)") }
        }
        fresh.onEvent = { [weak self] event in
            StreamEventBus.shared.emit(event)
            guard event.type == .failed else { return }
            DispatchQueue.main.async {
                self?.handleFailure(of: fresh, reason: event.reason ?? "reconnect exhausted")
            }
        }
    }

    private func handleStartFailed(isRestart: Bool, error: Error) {
        isStreamTransition = false
        if isRestart, isStreaming, !stopRequested, isReleased?() != true {
            scheduleRestart()
        } else {
            isStreaming = false
            stopRequested = false
            stopPathMonitor()
            if !isRestart {
                onMessage?("Не удалось начать стрим: \(error)", true)
            }
        }
        onStateChanged?()
    }

    private func handleStarted(_ fresh: StreamSession) {
        if isReleased?() == true || stopRequested {
            stopRequested = false
            isStreamTransition = false
            isStreaming = false
            stopPathMonitor()
            DispatchQueue.global(qos: .userInitiated).async { fresh.stop() }
            onStateChanged?()
            return
        }
        session = fresh
        isStreaming = true
        isStreamTransition = false
        onStateChanged?()
    }

    /// Сессия умерла (ядро прислало `failed` или энкодер выдал ошибку): отцепляем её и
    /// пересоздаём с backoff. Ошибки уже отцепленной сессии игнорируются.
    private func handleFailure(of failed: StreamSession, reason: String) {
        guard session === failed, isStreaming, isReleased?() != true else { return }
        session = nil
        DispatchQueue.global(qos: .userInitiated).async { failed.stop() }
        let now = Date()
        if lastErrorMessageAt.map({ now.timeIntervalSince($0) >= Self.errorMessageInterval }) ?? true {
            lastErrorMessageAt = now
            onMessage?("Ошибка стрима: \(reason). Переподключаемся…", true)
        }
        scheduleRestart()
    }

    // MARK: - Restart

    /// Backoff 2→4→8→15 с без лимита попыток.
    private func scheduleRestart() {
        guard restartTimer == nil, isReleased?() != true else { return }
        let delayMs = restartBackoff.nextDelayMs()
        NSLog("[mafbase_stream] stream restart #\(restartBackoff.attempt) in \(delayMs)ms")
        scheduleRestart(after: TimeInterval(delayMs) / 1000)
    }

    private func scheduleRestart(after delay: TimeInterval) {
        restartTimer = Timer.scheduledTimer(withTimeInterval: delay, repeats: false) { [weak self] _ in
            guard let self = self else { return }
            self.restartTimer = nil
            self.restartNow()
        }
    }

    private func cancelRestart() {
        restartTimer?.invalidate()
        restartTimer = nil
    }

    private func restartNow() {
        guard isStreaming, !isStreamTransition, isReleased?() != true else { return }
        if isRecordTransition?() == true {
            scheduleRestart(after: Self.transitionRetry)
            return
        }
        guard let fresh = makeSession.flatMap({ $0() }) else { return }
        let old = session
        session = nil
        isStreamTransition = true
        onStateChanged?()
        launch(fresh, replacing: old, isRestart: true)
    }

    // MARK: - Network path

    private func startPathMonitor() {
        guard pathMonitor == nil else { return }
        let monitor = NWPathMonitor()
        monitor.pathUpdateHandler = { [weak self] path in
            guard let self = self, self.pathMonitor === monitor else { return }
            let interfaces = path.availableInterfaces.map { "\($0.type)" }.sorted().joined(separator: ",")
            let key = "\(path.status)|\(interfaces)"
            let previous = self.lastPathKey
            self.lastPathKey = key
            // Первый колбэк приходит сразу при старте монитора — это не смена сети.
            guard previous != nil, previous != key, path.status == .satisfied else { return }
            self.handlePathChanged()
        }
        monitor.start(queue: .main)
        pathMonitor = monitor
    }

    private func stopPathMonitor() {
        pathMonitor?.cancel()
        pathMonitor = nil
        lastPathKey = nil
    }

    /// Сеть сменилась: если ждём backoff пересоздания или ядро в реконнекте — новая сессия
    /// сразу, старая после блипа обычно уже мёртвая.
    private func handlePathChanged() {
        guard isStreaming, isReleased?() != true else { return }
        let waitingForRestart = restartTimer != nil
        let coreReconnecting = session?.lastCoreState == .reconnecting
        guard waitingForRestart || coreReconnecting else { return }
        NSLog("[mafbase_stream] network path changed — reconnect now")
        cancelRestart()
        restartBackoff.reset()
        restartNow()
    }
}
